apps/aechannel/test/aesc_txs_SUITE.erl

```diff
@@ -2875,6 +2886,208 @@ fp_too_soon(Cfg) ->
+                                          _Deposit  = 2),
+                  % force progress contract on-chain
+                  fun(#{contract_id := ContractId,
+                        from_pubkey := Pubkey,
+                        from_privkey := Privkey} = Props) ->
+                      Sig = SignContractAddress(Pubkey, Privkey, ContractId),
+                      Account = aeu_hex:hexstring_encode(Pubkey),
+                      PreclaimArgs = <<"(", Account/binary, ",",
+                                            CHash/binary, ",",
+                                            Sig/binary,
+                                        ")">>,
+                      ?TEST_LOG("Off-chain preclaim args ~p", [PreclaimArgs]),
+                      (force_call_contract_first(Forcer, <<"preclaim">>,
+                                            PreclaimArgs, FPRound))(Props)
+                  end,
+                  % ensure all gas is consumed and channel is updated
+                  fun(#{state := S,
+                        channel_pubkey := ChannelPubKey,
+                        signed_force_progress := SignedForceProgressTx,
+                        solo_payload := #{update    := Update,
+                                        state_hash := ExpectedStateHash,
+                                        round      := ExpectedRound}} = Props) ->
+                      {_ContractId, Caller} = aesc_offchain_update:extract_call(Update),
+                      TxHashContractPubkey = aesc_utils:tx_hash_to_contract_pubkey(
+                                            aetx_sign:hash(SignedForceProgressTx)),
+                      CallId = aect_call:id(Caller,
+                                            FPRound,
+                                            TxHashContractPubkey),
+                      Call = aect_test_utils:get_call(TxHashContractPubkey, CallId,
+                                                      S),
+                      ?TEST_LOG("Off-chain call ~p", [Call]),
+                      {_, GasPrice, GasLimit} = aesc_offchain_update:extract_amounts(Update),
+                      %% assert all gas was consumed
+                      GasLimit = aect_call:gas_used(Call),
+                      GasPrice = aect_call:gas_price(Call),
+                      %% the default catch all reason for error
+                      <<"not_allowed_off_chain">> = aect_call:return_value(Call),
+                      error = aect_call:return_type(Call),
+
+                      %% expected channel states
+                      Channel = aesc_test_utils:get_channel(ChannelPubKey, S),
+                      FPRound = aesc_channels:round(Channel),
+                      FPRound = ExpectedRound,
+                      ExpectedStateHash = aesc_channels:state_hash(Channel),
+                      Props
+                  end])
+        end,
+    [Test(Owner, Forcer) || Owner  <- ?ROLES,
+                            Forcer <- ?ROLES],
+    ok.
```

Comment:
I see on-chain contracts are [not allowed to be called](https://github.com/aeternity/protocol/blame/26ba07353b62d236c3b9724f990479b8b4999e79/channels/OFF-CHAIN.md#L1053) and that implementation [ignores on-chain trees for calls](https://github.com/aeternity/epoch/blob/ee758a13d008134262bf83bd806e25b790840190/apps/aecore/src/aec_vm_chain.erl#L443) that seems ok. Though I see no test for that (out of scope for this PR). Is it tracked anywhere to eventually write one?

Reply:
Tracked as [SC contract: Call on-chain test (negative test)](https://www.pivotaltracker.com/story/show/161215899)